### PR TITLE
[FRR]: Port some patches from sonic-quagga repo

### PR DESCRIPTION
**Pull request by pavel-shirshov**
**Please don't merge it**

**- What I did**
Port 3 patches from sonic-quagga.

**- How I did it**
Manually

**- How to verify it**
Build it and run on your dut
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by lguohan** [comment]
can you uncomment the invalid_nexthop vs test to make the patch is working?

**Comment by pavel-shirshov** [comment]
@lguohan Sure. Already did that.

**Comment by lguohan** [comment]
vs test failed. please check. https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-all-pr/85/console

**Comment by nikos-github** [comment]
The nexthop patch is not correct. I have pushed a different version of it a while back into frr latest and I also have diffs to push for the others. Best to hold on to these for a bit until I push into frr the rest of the code.

**Comment by pavel-shirshov** [comment]
@nikos-github Thank you for your comment. I've ported your patch from the master to frr-7.
What do you mean by "The log change needs to be under a cli."?
Thanks

**Comment by pavel-shirshov** [comment]
retest this please

**Comment by pavel-shirshov** [comment]
retest vs please

**Comment by nikos-github** [comment]
> @nikos-github Thank you for your comment. I've ported your patch from the master to frr-7.
> What do you mean by "The log change needs to be under a cli."?
> Thanks

The change from info to debug in vty log needs to go under a cli cmd option. I have that already.

Also the other patch that Guohan pointed me in quagga regarding the dscp, is also not correct and I had pointed that out to him and I have fixed in my diffs.

**Comment by pavel-shirshov** [comment]
@nikos-github 
Where can I find your patches?
Thanks

File: src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+diff --git a/bgpd/bgp_route.c b/bgpd/bgp_route.c
+index 38f3cad5a..55240eab8 100644
+--- a/bgpd/bgp_route.c
++++ b/bgpd/bgp_route.c
+@@ -2873,8 +2873,7 @@ static int bgp_update_martian_nexthop(struct bgp *bgp, afi_t afi, safi_t safi,
+ 
+ 	/* If NEXT_HOP is present, validate it. */
+ 	if (attr->flag & ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP)) {
+-		if (attr->nexthop.s_addr == 0
```
Comment (nikos-github) [inline src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: What failure are you seeing with the presence of the previous patch above that's forcing you to make this change?

**Comment by lguohan** [comment]
retest vs please

**Comment by lguohan** [comment]
retest vs please

**Comment by lguohan** [comment]
@pavel-shirshov , tests are still failing.

File: src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch (hunk 1 above)
Comment (pavel-shirshov) [reply src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: https://github.com/Azure/sonic-buildimage/blob/master/platform/vs/tests/bgp/test_invalid_nexthop.py to pass the test

Comment (nikos-github) [reply src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: @pavel-shirshov @lguohan The test is wrong.

Comment (nikos-github) [reply src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: @pavel-shirshov @lguohan This isn't the right way to address the case of an IPv6 NLRI being sent and a nexthop attribute of 0. The flag needs to be cleared earlier. The diffs are just punching a hole in the code but the nexthop attribute flag still remains set.

Comment (paulnice) [reply src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: @nikos-github Currently, when we allow bgpd to have 0.0.0.0 as next-hop ip address, bgpd tries to resolve the next-hop using zebra, zebra responds with the message "unresolved" and this next-hop is not used.
Also, why do you consider that this check was valid? I tried to find in the RFC why NEXT_HOP attribute can't be 0.0.0.0 and I didn't find anything. Probably I miss something. Can you please point me to the place in the RFC which says us we should filter-out next_hops equal to 0.0.0.0?

Comment (nikos-github) [reply src/sonic-frr/patch/0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch]: @paulnice There is no such thing as nexthop of 0.0.0.0 for a prefix learned via a neighbor. Nexthops for prefixes received by peers, correspond to an interface address. 0.0.0.0 is not a valid interface address. It's all described in the RFC.

**Comment by imzyxwvu** [comment]
Hello, could someone tell me what vendor bug is `0004-Allow-BGP-attr-NEXT_HOP-to-be-0.0.0.0-due-to-allevia.patch` trying to alleviate. Is there any branded switch sending route update with the Next Hop property value 0.0.0.0?

**Comment by nikos-github** [comment]
Arista.

**Pull request merged**